CUDA support in the CSR layout:  CUDA addmm/matvec

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59012 CUDA support in the CSR layout:  CUDA addmm/matvec**

Restack of #57403, see review history there.



Differential Revision: [D28719631](https://our.internmc.facebook.com/intern/diff/D28719631)

aocsa committed:
CUDA support in the CSR layout:  CUDA addmm/matvec

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28719631).

bhosmer reviewed (approved):
Transferring approvals from https://github.com/pytorch/pytorch/pull/57403

ngimel commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -49,68 +51,74 @@ namespace {
+        }
+        Tensor dense_;
+        char transpose_dense;
+        if(dense.stride(0) == 1 && dense.stride(1) == dense.size(0)) {
```
given that cusparse APIs support setting ldb, this seems like a too strict condition? dense.stride(1) can be larger than dense.size(0)

aocsa replied:
Thanks @ngimel , currently `ldb` is set in this way: https://github.com/pytorch/pytorch/blob/1bd22e28b318b15cbfe37ee1a112df9423aab328/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu#L106

If I understand well the first stride `stride(1)` can be `size(0) * s` where `s` is some positive integer.  I could compute `s`  to consider this case, however  I would like to test this too, but I'm not sure how can I generate this test case.  cc @pearu

pearu replied:
Indeed, given that csrmm2 supports strided inputs, forcing `dense` to be contiguous is unnecessary.
In fact, the same applies to `r`.

@ngimel, given that @aocsa has only 3 days left to work on this ghstack, would it be ok to land this PR provided that @aocsa will create the corresponding issues that we'll address in follow-up PRs?

aocsa replied:
cc @bhosmer

ngimel replied:
Sure, we can file an issue about relaxing constraints and not pursue it here. I would like to better understand what's going on with device checking though.

ngimel commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -33,6 +35,74 @@ using namespace at::sparse_csr;
+
+  TORCH_INTERNAL_ASSERT(t.device().type() == kCUDA);
+  TORCH_CHECK(
+      r.device().type() == kCUDA,
```
why are you checking only device types and not exact devices? Also, likely the device checks are not needed at all because they are now automatically generated (but you should make sure that this is tested)

aocsa replied:
You are right, I tested it and these checks are automatically generated thanks to https://github.com/pytorch/pytorch/pull/56872. I will remove them as they are redundant.

aocsa replied:
Update: check are automatically generated but is not always called. I think this should be addressed in a different PR as it not only related to CSR tensors but COO tensors as well. 
I detailed the issue here: https://github.com/pytorch/pytorch/issues/59058

ngimel commented on test/test_sparse_csr.py:
```diff
@@ -371,33 +369,48 @@ def test_csr_matvec(self, device, dtype):
+                expected = torch.addmm(t, x.to_dense(), y)
+                self.assertEqual(res, expected)
+
+                res = torch.mm(x, y)
```
add tests that make sure error is raised with device mismatch. Also, various discontiguous and/or broadcasted dense inputs should be tested

aocsa replied:
Added: Test with discontinuous and/or broadcasted dense inputs. 

Can't add: Tests that make sure error is raised with device mismatch. This neither works with COO tensors. I create an issue for this here:   #59058

ngimel replied:
I'm confused, you have device checks in your implementation - do they not throw an error?

aocsa replied:
Yes, It throws an error. That's why I maintained the device check in this implementation. 
What I am saying is If I disabled these checks  and trust in the 
automatically generated checks it will fail as it reported here:   #59058

ngimel replied:
So you can actually add a test that assertRaises on device mismatch? If autogeneration is not working, then please don't disable the checks in your implmenetation, and add tests, so that when autogeneration is fixed it can easily be tested.

aocsa replied:
Ok I will add these test with current checks.

aocsa committed:
minor update/feedback comments addressed on "CUDA support in the CSR layout:  CUDA addmm/matvec"


Restack of #57403, see review history there.

Differential Revision: [D28719631](https://our.internmc.facebook.com/intern/diff/D28719631)

[ghstack-poisoned]

aocsa committed:
add test_matmul_device_mismatch on "CUDA support in the CSR layout:  CUDA addmm/matvec"


Restack of #57403, see review history there.



Differential Revision: [D28719631](https://our.internmc.facebook.com/intern/diff/D28719631)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28719631).

aocsa committed:
minor fix test_matmul_device_mismatch on "CUDA support in the CSR layout:  CUDA addmm/matvec"


Restack of #57403, see review history there.



Differential Revision: [D28719631](https://our.internmc.facebook.com/intern/diff/D28719631)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28719631).

aocsa committed:
rebased needed due to merge conflicts on "CUDA support in the CSR layout:  CUDA addmm/matvec"


Restack of #57403, see review history there.



Differential Revision: [D28719631](https://our.internmc.facebook.com/intern/diff/D28719631)

[ghstack-poisoned]

aocsa commented:
> @bhosmer has imported this pull request. If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28719631).

I rebased this PR due to merge conflicts reported in the CI. I think it will need to re-import to finally merge. cc  @bhosmer

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28719631).
